docs/guides/spmd_api.md

```diff
@@ -0,0 +1,429 @@
+state_spec = nn.get_partition_spec(abstract_variables)
+state_spec
+```
+
+<!-- #region id="KAyPfNP9pTR4" -->
+# `pjit` the initialization and train step
+
+Now you can `pjit` the `init_fn` in a similar fashion as `jit`, with two extra args `in_axis_resources` and `out_axis_resources`.
+
+You need to add `with mesh:` context when running a `pjit`ted function, so that it can refer to `mesh` to allocate data on devices correctly.
+<!-- #endregion -->
+
+```python id="RFxvZjwPS5Za"
+pjit_init_fn = pjit(init_fn,
+                    static_argnums=(2, 3),
+                    in_axis_resources=(PartitionSpec(None), x_spec),  # RNG key and x
+                    out_axis_resources=state_spec
+                    )
+with mesh:
+  initialized_state = pjit_init_fn(k, x, model, optimizer)
+jax.tree_map(jnp.shape, initialized_state)
+```
+
+<!-- #region id="qgVa3lqLgPLz" -->
+## Looking into the module output
+
+Note that in the output `initialized_state`, the params `W1` and `W2` are of type [`Partitioned`](https://flax.readthedocs.io/en/latest/api_reference/_autosummary/flax.linen.Partitioned.html). This is a wrapper around the actual JAX array that allows Flax to record metadata associated with it. You can access the raw JAX array by adding `.value` or running `.unbox()`.
+
+You can also check the underlying `.sharding` of the JAX array, which gives a hint on the way it is partitioned.
+<!-- #endregion -->
+
+```python id="UEj1A2BhhUcC"
+print(type(initialized_state.params['ScanDotReluDot_0']['W1']))
+print(type(initialized_state.params['ScanDotReluDot_0']['W1'].value))
+print(initialized_state.params['ScanDotReluDot_0']['W1'].value.shape)
+```
+
+```python id="-Hhp_glBOL1R"
+print(initialized_state.params['ScanDotReluDot_0']['W1'].value.sharding)
+```
+
+<!-- #region id="Yu7ohrBzN-eP" -->
+You can use `jax.tree_map` to perform mass computation on a dict of boxed params, in the same way as on a dict of JAX arrays.
+<!-- #endregion -->
+
+```python id="aqwt7F0Kmn0f"
+diff = jax.tree_map(
+    lambda a, b: a - b, 
+    initialized_state.params['ScanDotReluDot_0'], initialized_state.params['ScanDotReluDot_0'])
+print(jax.tree_map(jnp.shape, diff))
+diff_array = diff['W1'].unbox()
+print(type(diff_array))
+print(diff_array.shape)
+```
+
+<!-- #region id="dcLFwH1dhr3v" -->
+## `pjit` the train step and inference 
+
+Now we can do the same and make a `pjit`ted training step.
+<!-- #endregion -->
+
+```python id="KF5lmJ5FwxqJ"
+def train_step(state, x):
+  # fake loss fn
+  def loss_unrolled(params):
+    y = model.apply({'params': params}, x)
+    return y.sum()
+  grad_fn = jax.grad(loss_unrolled)
+  grads = grad_fn(state.params)
+  state = state.apply_gradients(grads=grads)
+  return state
+
+pjit_step_fn = pjit(train_step,
+                    (state_spec, x_spec),  # input annotations
+                    state_spec,            # output annotations
+                    )
+```
+
+```python id="F5gy_gb_mOqR"
+with mesh:
```

Comment:
nit: in the other examples, the `with mesh:` code is in the same cell as the function being pjitted